flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1481,6 +1481,17 @@ class ScalarExprLowering {
           delta = builder.create<mlir::MulIOp>(loc, delta, ext);
         ++dim;
       }
+      if (fir::factory::CharacterExprHelper::isCharacterScalar(refTy)) {
+        auto chTy = fir::factory::CharacterExprHelper::getCharacterType(refTy);
+        if (fir::characterWithDynamicLen(chTy)) {
+          auto ctx = builder.getContext();
+          auto kind = fir::factory::CharacterExprHelper::getCharacterKind(chTy);
+          auto singleTy = fir::CharacterType::getSingleton(ctx, kind);
+          refTy = builder.getRefType(singleTy);
+          auto seqRefTy = builder.getRefType(builder.getVarLenSeqTy(singleTy));
+          base = builder.createConvert(loc, seqRefTy, base);
+        }
+      }
       return builder.create<fir::CoordinateOp>(
           loc, refTy, base, llvm::ArrayRef<mlir::Value>{total});
```

Comment:
I am wondering if there is a risk that later uses of the `fir.coordinate_of` result think this is a `CHARACTER(1)` looking at the address type (In the test below, the consumers are casting to fir.char<kind, ?> because we are always creating embox char from fir.char<kind, ?>). Wouldn't it be safer to cast back the result type to `fir.char<kind, ?>` directly here ?

Reply:
I can do that.